pkg/controller/core.oam.dev/v1alpha2/application/application_controller.go

```diff
@@ -211,53 +218,45 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
-		if duration := wf.GetSuspendBackoffWaitTime(); duration > 0 {
-			_, err = r.gcResourceTrackers(logCtx, handler, common.ApplicationWorkflowSuspending, false, true)
+		if duration := executor.GetSuspendBackoffWaitTime(); duration > 0 {
+			_, err = r.gcResourceTrackers(logCtx, handler, common.ApplicationWorkflowSuspending, false, isUpdate)
 			return r.result(err).requeue(duration).ret()
 		}
 		if !workflow.IsFailedAfterRetry(app) || !feature.DefaultMutableFeatureGate.Enabled(features.EnableSuspendOnFailure) {
 			r.stateKeep(logCtx, handler, app)
 		}
-		return r.gcResourceTrackers(logCtx, handler, common.ApplicationWorkflowSuspending, false, true)
-	case common.WorkflowStateTerminated:
+		return r.gcResourceTrackers(logCtx, handler, common.ApplicationWorkflowSuspending, false, isUpdate)
+	case workflowv1alpha1.WorkflowStateTerminated:
 		logCtx.Info("Workflow return state=Terminated")
 		handler.UpdateApplicationRevisionStatus(logCtx, handler.latestAppRev, false, app.Status.Workflow)
-		if err := r.doWorkflowFinish(app, wf, workflowState); err != nil {
-			return r.endWithNegativeCondition(ctx, app, condition.ErrorCondition(common.WorkflowCondition.String(), errors.WithMessage(err, "DoWorkflowFinish")), common.ApplicationRunningWorkflow)
-		}
-		return r.gcResourceTrackers(logCtx, handler, common.ApplicationWorkflowTerminated, false, true)
-	case common.WorkflowStateExecuting:
+		r.doWorkflowFinish(app, workflowState)
+		return r.gcResourceTrackers(logCtx, handler, common.ApplicationWorkflowTerminated, false, isUpdate)
+	case workflowv1alpha1.WorkflowStateExecuting:
 		logCtx.Info("Workflow return state=Executing")
-		_, err = r.gcResourceTrackers(logCtx, handler, common.ApplicationRunningWorkflow, false, true)
-		return r.result(err).requeue(wf.GetBackoffWaitTime()).ret()
-	case common.WorkflowStateSucceeded:
+		_, err = r.gcResourceTrackers(logCtx, handler, common.ApplicationRunningWorkflow, false, isUpdate)
+		return r.result(err).requeue(executor.GetBackoffWaitTime()).ret()
+	case workflowv1alpha1.WorkflowStateSucceeded:
 		logCtx.Info("Workflow return state=Succeeded")
 		handler.UpdateApplicationRevisionStatus(logCtx, handler.currentAppRev, true, app.Status.Workflow)
-		if err := r.doWorkflowFinish(app, wf, workflowState); err != nil {
-			return r.endWithNegativeCondition(logCtx, app, condition.ErrorCondition(common.WorkflowCondition.String(), errors.WithMessage(err, "DoWorkflowFinish")), common.ApplicationRunningWorkflow)
-		}
+		r.doWorkflowFinish(app, workflowState)
 		app.Status.SetConditions(condition.ReadyCondition(common.WorkflowCondition.String()))
 		r.Recorder.Event(app, event.Normal(velatypes.ReasonApplied, velatypes.MessageWorkflowFinished))
 		logCtx.Info("Application manifests has applied by workflow successfully")
 		if !EnableReconcileLoopReduction {
-			return r.gcResourceTrackers(logCtx, handler, common.ApplicationWorkflowFinished, false, true)
-		}
-	case common.WorkflowStateFinished:
```

Comment:
why this case removed？

Reply:
There's no redundant finished state in workflow anymore.